Core utility

dwalsten5 opened the pull request:
This is my first attempt at defining the core functionality of the storageUtility class. I implemented a tier-based invalidation approach and wrote a test suite to verify that it functions as expected.

To test:
- [ ] Please review my source code and let me think of my implementation.
- [ ] Check out my test suite and make sure that I'm covering all of my bases. There is an AWESOME jest package that lets me mock either sessionStorage or localStorage.

Let me know if you have any questions! I was also wondering if we should release this as a private package on npm first while we tinker with it rather than immediately publish it publicly?

dwalsten5 committed:
Basic setup!

dwalsten5 committed:
Add isAvailable functionality

Now, we need to figure out how to test this out.

dwalsten5 committed:
Add more thorough testing

dwalsten5 committed:
More adjustments

dwalsten5 committed:
Rough draft of README.

dwalsten5 committed:
Add missing javascript tag

dwalsten5 committed:
Remove tab

dwalsten5 committed:
Whoops

dwalsten5 committed:
Define remove, removeAll methods

dwalsten5 committed:
Add tests for removing

dwalsten5 committed:
Add more clarification comments

dwalsten5 committed:
Exclude tiers with whitespace

dwalsten5 committed:
Add some missing dependencies

dwalsten5 committed:
Use target to store last-change, update tests

This allows us to invalidate whenever the page refreshes.

nktasingh commented on src/StorageUtility.js:
```diff
@@ -1,11 +1,195 @@
+	static isAvailable(type) {
+		// Helper method to determine if both the store exists and has space
+		const storage = window[type];
+		try {
```
I like this use of a try/catch!!

nktasingh commented on src/StorageUtility.js:
```diff
@@ -1,11 +1,195 @@
+			.filter((tier) => {
+				// We don't want any tiers with white space in the name
+				if (tier.name.match(/\s/g)) {
+					console.error(`Invalid tier name "${tier.name}" with whitespace`);
```
Should we explain this restriction in the readme too? Great docs here!

dwalsten5 replied:
Good call, I will add this to the Readme

nktasingh commented on src/StorageUtility.js:
```diff
@@ -1,11 +1,195 @@
+	}
+
+	// Helper method to determine if both the store exists and has space
+	isAvailable() {
```
Is there a point to declaring this both as a static and class method?

dwalsten5 replied:
Yeah, I wasn't sure myself. I was thinking that there might be a case where the user wants to check whether the cache is available even before initializing an instance of StorageUtility.

nktasingh commented on src/StorageUtility.js:
```diff
@@ -1,11 +1,195 @@
 class StorageUtility {
```
The algorithms going on here are intricate enough that I think it's worth you walking me through this!! Maybe we can do that during our sprint meeting and include Steph as well?

dwalsten5 replied:
Makes sense to me! I'd be more than happy to walk through it.

dwalsten5 committed:
Include comment on whitespace in README

dwalsten5 committed:
More limiting check of prefix

Actually verify that the filter is at the start of the string. We may to consider a hash function to reduce the risk of interfering with other values in local storage.

dwalsten5 committed:
Unique keys

dwalsten5 committed:
Prevent user from including tier name in key

dwalsten5 committed:
Properly remove keys from keyMap on removeAll

dwalsten5 committed:
Restructure invalidate

Now, we just return a bool that says whether invalidation was successful. In invalidation is unsuccessful, do not get or set. However, we can still remove.

dwalsten5 commented:
Thanks for the feedback! This is good for another review whenever you're ready.

dwalsten5 committed:
Update name, license in package.json

dwalsten5 committed:
Create LICENSE

dwalsten5 committed:
Modify build processs, names, try t setup Travis

dwalsten5 commented:
Alrigh @nktasingh , this is ready for another review whenever you are ready.

dwalsten5 committed:
Add clean script

dwalsten5 commented:
Shoot, I realize now another reason why I needed the user to supply the tier name on a `get`. If the user refreshes the page, we lose the utility instance and the mapping of keys in storage. This means that I will have to parse the existing storage object and identify any keys that match all possible prefixes based on the tier config. I think this is a low cost operation at startup for the benefit of a simpler API, but I wanted to get your opinion.

dwalsten5 commented:
@schew in @nktasingh's absence, if you could give this a final review, I would greatly appreciate it!

schew commented on __tests__/core.test.js:
```diff
@@ -0,0 +1,211 @@
+		expect(console.error).toHaveBeenCalledWith('Invalid tier name "D Tier" with whitespace');
+	});
+
+	it('should properly set a value in target', () => {
```
what should happen when we set this:
```
{
   hello: '123',
   goodbye: '456',
}
```

dwalsten5 replied:
That whole object should be stored at the key `test` in storage.

schew commented on README.md:
```diff
@@ -1,2 +1,57 @@
+Frontend caching has many benefits. One of the biggest improvements it can provide is to reduce the number of requests made to the backend by storing data that rarely changes on the frontend. This reduces the time the user spends waiting for data to load on a particular page and leads to a more responsive application. The `sessionStorge` and `localStorage` properties on the browser window provide ...
+
+## Approach
+storageUtility provides a higher-level API for working with these storage entities. It uses a tier-based approach to invalidating data in the cache. The user can specify tiers within the cache with specific expiration times. If the storage entity has been inactive for a greater amount of time than this expiration, all keys within this tier are removed from storage. This is a passive approach to m ...
```
are we renaming this to cacheUtility? i think i've seen both, so just want to make sure i'm on the same page :P

dwalsten5 replied:
Gah! Thanks for catching this in the README. Yes, we are going to use cache-utility.

schew commented on src/CacheUtility.js:
```diff
@@ -0,0 +1,254 @@
+			}
+		});
+
+		if (this.isAvailable()) {
```
in continuation with nikita's comment about the two functions here, is there a reason why we want to use `this.isAvailable` rather than `CacheUtility.isAvailable(config.target)`?

dwalsten5 replied:
Technically, that is valid, too. I guess in my mind, I only wanted to have the static method used if the end user is trying to decide whether or not they even want to create an instance of CacheUtility. I wanted to have the isAvailable method tied to the instance of CacheUtility so I could use the instance to determine whether the cache is available rather than import CacheUtility just to have access to this method.

dwalsten5 replied:
I would not be opposed to rewriting the instance method to use the static method to reduce on code duplication if that helps?

schew reviewed (commented):
Wowowowoow it looks really clean and very straightforward to use. I also like how you include examples with the docs so it's pretty clear what you're getting out of this

dwalsten5 committed:
Use static method in isAvailable

Code reuse

dwalsten5 committed:
Rename everything to cache-utility

whoops

dwalsten5 commented:
Thanks for the review @schew! This is ready for another review whenever you're ready.

dwalsten5 committed:
Rename test

schew commented:
looks good!

dwalsten5 committed:
Update snapshots
